README.md
# offbeatSlider - a jQuery Slideshow Plugin

Easy to use jQuery plugin to create an image slideshow for a thumbnail preview or
for page headers.

**Version: 0.9.2**

## Installation

You can download the latest release [here](https://github.com/PioBeat/offbeatSlider/releases).
Just copy the [css](https://raw.githubusercontent.com/PioBeat/offbeatSlider/master/css/offbeat-slider.css) and 
[js](https://raw.githubusercontent.com/PioBeat/offbeatSlider/master/js/offbeat-slider.js) file into your project directory.
Then include it like this in your web page:
```html
<!doctype html>
<html lang="">
<head>
    <!-- YOUR HEADER -->
    <link rel="stylesheet" href="css/offbeat-slider.css">
</head>
<body>

    <!-- YOUR CONTENT -->

    <!-- jQuery -->
    <script src="https://code.jquery.com/jquery-1.12.0.min.js"></script>

    <!-- offbeatSlider Plugin -->
    <script src="js/offbeat-slider.js"></script>
</body>
</html>
```

## Quick Example

This example shows two different styles for the positioning of the navigation controls and the setting of the start image:

<img src="/art/example1.png" width="49%" />

Code:
```html
<!-- Left example -->

<div class="ofp-slider" data-slider-index="1">
    <img class="ofp-slides" src="img/Hydrangeas.jpg">
    <img class="ofp-slides" src="img/Jellyfish.jpg">
    <img class="ofp-slides" src="img/Koala.jpg">

    <div class="ofp-slider-navigation ofp-center ofp-middle">
        <div class="ofp-arrow-left">&#10094;</div>
        <div class="ofp-arrow-right">&#10095;</div>
    </div>
    <div class="ofp-slider-navigation ofp-bottom-middle">
        <span class="ofp-slider-dots"></span>
        <span class="ofp-slider-dots"></span>
        <span class="ofp-slider-dots"></span>
    </div>
</div>


 <!-- Right example -->

<div class="ofp-slider" data-slider-index="2">
    <img class="ofp-slides" src="img/Hydrangeas.jpg">
    <img class="ofp-slides" src="img/Jellyfish.jpg">
    <img class="ofp-slides" src="img/Koala.jpg">

    <div class="ofp-slider-navigation ofp-top-right">
        <div class="ofp-arrow-left"> &#10096;</div>
        <div class="ofp-arrow-right"> &#10097;</div>
    </div>
    <div class="ofp-slider-navigation ofp-center ofp-bottom-middle">
        <span class="ofp-slider-dots"></span>
        <span class="ofp-slider-dots"></span>
        <span class="ofp-slider-dots"></span>
    </div>
</div>
```

## Basic Usage

### HTML
To transform a div into a slide show add the class ``ofp-slider`` to a div and append some images as direct child. For each image
add the class ``ofp-slides``.
```html
<div class="ofp-slider" data-slider-index="1">
    <img class="ofp-slides" src="img/Hydrangeas.jpg">
    <img class="ofp-slides" src="img/Jellyfish.jpg">
    <img class="ofp-slides" src="img/Koala.jpg">

    <div class="ofp-slider-navigation ofp-center ofp-middle">
        <div class="ofp-arrow-left">&#10094;</div>
        <div class="ofp-arrow-right">&#10095;</div>
    </div>
    <div class="ofp-slider-navigation ofp-bottom-middle">
        <span class="ofp-slider-dots"></span>
        <span class="ofp-slider-dots"></span>
        <span class="ofp-slider-dots"></span>
    </div>
</div>
```
The data attribute ``slider-index`` specifies the index of the first slide to show.

### JavaScript
Next, put this piece of code at the end of your page:
```javascript
$(".ofp-slider").offbeatSlider({"slideStartIndex": 1});
```
The option ``slideStartIndex`` defines the index of the first slide to show for every slide show div if
the data attribute ``slider-index`` is not defined.

## Options

**First slide** (JavaScript / data-attribute)

JavaScript property ``slideStartIndex`` sets the first slide image for all slide show containers if no
data-attribute ``slider-index`` is specified in the corresponding div with the class ``ofp-slider``.

**Animation**

You can set the transition animation for each slide in the slide show with these options:
```javascript
$(".ofp-slider").offbeatSlider({
                                   animate: true,
                                   duration: 1000,
                                   animation: "normal",
                                   easing: "linear" //easein, linear, ...
                               });
```
You can adjust the duration and the easing function. The name for the easing functions are the same as in jQuery.
This is the default slider implementation. A simple and beautiful fade-in transition is created.

If you want to use specific CSS3 animations you have to declare the following settings:
```javascript
$(".ofp-slider").offbeatSlider({
                                   animate: true,
                                   animation: "css",
                                   animationRight: "swing-in-right-fwd",
                                   animationLeft: "swing-in-left-fwd"
                               });
```

The ``animation`` property tells the slider to use the ``animationRight`` and ``animationLeft`` values which are CSS animations
for both directions. Those classes have to be available in a css file. The same settings can also be made available in the correspondig div
with the data-attributes ``data-animation-right`` and ``data-animation-left``:
```html
<div class="ofp-slider" data-animation-right="swing-in-right-fwd" data-animation-left="swing-in-left-fwd">
<!-- more ... -->
</div>           
```

To create CSS animations you can use [animista](http://animista.net).

**Carousel**

Automatically show the available slides in sequence with these options:
```javascript
$(".ofp-slider").offbeatSlider({
            carousel: true, //auto-animate ?
            carouselDelay: 3000
        });
```

The default of the carousel animation is ``false`` which means the slides won't be shown automatically one by one.

## Customisation

### Positioning of navigation controls

**Combined**
Slider dots and arrows as one unit:
```html
<div class="ofp-slider-navigation ofp-center ofp-middle">
    <div class="ofp-arrow-left">&#10094;</div>
    <div class="ofp-arrow-right">&#10095;</div>
    <span class="ofp-slider-dots"></span>
    <span class="ofp-slider-dots"></span>
    <span class="ofp-slider-dots"></span>
</div>
```
Both controls will be displayed centered horizontally and vertically within the image.

**Separated**
Use two div-containers with class ``ofp-slider-navigation``:
```html
<div class="ofp-slider-navigation ofp-center ofp-top-middle">
    <div class="ofp-arrow-left">&#10094;</div>
    <div class="ofp-arrow-right">&#10095;</div>
</div>

<div class="ofp-slider-navigation ofp-center ofp-bottom-middle">
    <span class="ofp-slider-dots"></span>
    <span class="ofp-slider-dots"></span>
    <span class="ofp-slider-dots"></span>
</div>
```
Arrow controls will be displayed at the top and the dots are displayed at the bottom within the image.

#### <a name="positioning"></a> Classes for positioning

Any meaningful combination of the listed css classes is possible to position
the navigation controls or the image description.

```css
ofp-center
ofp-top-left
ofp-top-right
ofp-bottom-left
ofp-bottom-right
ofp-middle
ofp-left
ofp-right
ofp-top-middle
ofp-bottom-middle
```

### Text slide

You can have a text-only slide if you add the css class ``ofp-slides-text`` next to ``ofp-slides``:
```html
<div class="ofp-slider" data-slider-index="1">
    <div class="ofp-slides ofp-slides-text">
        I'm a nice text slide
    </div>
    <img class="ofp-slides" src="img/Hydrangeas.jpg">
    <div class="ofp-slides ofp-slides-text">
        I'm the <strong>last</strong> nice text slide
    </div>
    <div class="ofp-slider-navigation ofp-top-right">
        <div class="ofp-arrow-left">&#10096;</div>
        <div class="ofp-arrow-right">&#10097;</div>
    </div>
</div>
```
With this piece of code you have 3 slides consisting of a text slide, image slide, and a text slide. The
navigation is placed at the top right corner.

### Image slide with description
For that case:
```html
<div class="ofp-slides">
    <img src="img/Hydrangeas.jpg">
    <p class="ofp-top-middle" style="color: #efefef; font-size: 1em; font-weight: bold;">
        Description for the picture
    </p>
</div>
```
For the positioning of the description use the same classes as described in <a href="#positioning">Classes for positioning</a>

### Arrow symbols

Use any character within the div-container with class ``ofp-arrow-left`` to represent the left arrow and with
class ``ofp-arrow-right`` for the right arrow.

To change the _size of the symbols_ add the ``font-size`` property to the mentioned classes.

**Examples**

The following table shows some common arrow symbols with their corresponding HTML codes:

HTML code| Symbol
---------|---------
``&#10094;``|&#10094;
``&#10095;``|&#10095;
``&#10096``|&#10096;
``&#10097``|&#10097;
``&#171;``|&#171;
``&#187;``|&#187;

### Further examples

Please look into the [test page](index.html) for further examples.

## Websites using <i>offbeatSlider</i>

<p>feel free to send me your website using this component! :)</p>

* [DollStuff](http://www.dollstuff.de/)

Changelog.md
# Changelog

## [0.9.2] - 2017-10-22
- Support for CSS animations for slide transition
- Fixed issue regarding the settings object. Settings are now shared correctly among the sliders.
Individual settings now possible for single sliders.
- Some code improvements

## [0.9.1.1] - 2017-04-27
Minor CSS update - Adjustment of arrow and slider dot size when
showing on a smaller screen

## [0.9.1] - 2017-03-29
- Possibility to have text-only slides
- Readme updated to show usage of an image slide with description

## [0.9] - 2017-03-26

### Added
- Transition animation to next slide via JavaScript options object
- Auto carousel
- Code refactoring and improvements

## [0.5] - 2017-03-26
Initial release

js/offbeat-slider.js
/**
 * jQuery Plugin offbeatSlider
 *
 * Version 0.9.2
 *
 * @author Dominik Grzelak
 * @since 2017-03-26
 */
(function ($) {

    var DATA_PARAM_INDEX = "slider-index";
    var DATA_PARAM_ANIMATION_RIGHT = "animation-right";
    var DATA_PARAM_ANIMATION_LEFT = "animation-left";

    var sessionObject = {};

    function createID() {
        return Math.ceil((Math.random()) * 0x100000).toString();
    }

    $.fn.offbeatSlider = function (options) {
        this.filter(".ofp-slider").each(function () {
            var sliderContainer = $(this);
            var uid = createID();
            var settings = $.extend({}, $.fn.offbeatSlider.settingsDefault, options);

            var tmpIx = sliderContainer.data(DATA_PARAM_INDEX);
            if (tmpIx !== undefined) {
                settings.slideStartIndex = tmpIx;
            }
            var animR = sliderContainer.data(DATA_PARAM_ANIMATION_RIGHT);
            var animL = sliderContainer.data(DATA_PARAM_ANIMATION_LEFT);
            if (animR !== undefined && animL !== undefined) {
                settings.animationRight = animR;
                settings.animationLeft = animL;
            }

            sessionObject[uid] = {
                "sliderContainer": sliderContainer,
                "settings": settings
            };
            init(uid);
        });
    };

    function init(uid) {
        var sliderContainer = sessionObject[uid].sliderContainer;
        var settings = sessionObject[uid].settings;
        var dots = sliderContainer.find(".ofp-slider-dots");
        var slides = sliderContainer.find(".ofp-slides");

        sessionObject[uid].stopCarousel = false;
        sessionObject[uid]["dots"] = dots;
        sessionObject[uid]["slides"] = slides;
        sessionObject[uid]["numSlides"] = slides.length;
        sessionObject[uid]["currentIndex"] = settings.slideStartIndex;

        for (var i = 0; i < dots.length; i++) {
            $(dots[i]).removeClass("ofp-slider-dots-active");
            $(dots[i]).click({uid: uid, index: i + 1}, clLis);
        }

        var left = sliderContainer.find(".ofp-arrow-left");
        var right = sliderContainer.find(".ofp-arrow-right");
        left.click({uid: uid}, clLis);
        right.click({uid: uid}, clLis);

        if (slides.length === 0) {
            return;
        }

        if (settings.carousel) {
            $(sliderContainer).mouseover(function () {
                sessionObject[uid].stopCarousel = true;
                clearTimeout(sessionObject[uid].carouselTimer);
            });
            $(sliderContainer).mouseout(function () {
                sessionObject[uid].stopCarousel = false;
                carousel(uid);
            });
            showNextSlide(uid);
            carousel(uid);
        } else {
            showNextSlide(uid);
        }
    }

    var clLis = function (event) {
        var uid = event.data.uid;
        var sliderContainer = sessionObject[uid].sliderContainer;
        var index = 0;
        if (event.data.index !== undefined) {
            index = event.data.index;
        } else {
            index = sliderContainer.data("slider-index");
        }

        sessionObject[uid].prevIndex = index;

        if (!$(event.target).hasClass("ofp-slider-dots")) {
            if ($(event.target).hasClass("ofp-arrow-left")) {
                index += -1;
            } else {
                index += 1;
            }
        }

        sessionObject[uid].currentIndex = index;
        showNextSlide(uid);
    };

    function carousel(uid) {
        if (sessionObject[uid].stopCarousel) {
            clearTimeout(sessionObject[uid].carouselTimer);
        }
        var settings = sessionObject[uid].settings;
        sessionObject[uid].carouselTimer = setTimeout(function () {
            var index = sessionObject[uid].currentIndex;
            var numSlides = sessionObject[uid].numSlides;
            index = checkSlideIndex(index + 1, numSlides);
            sessionObject[uid].currentIndex = index;
            showNextSlide(uid);
            carousel(uid);
        }, settings.carouselDelay);
    }

    function checkSlideIndex(index, maximum) {
        if (index === undefined || index > maximum) index = 1;
        if (index < 1) {
            index = maximum;
        }
        return index;
    }

    function showNextSlide(uid) {
        var settings = sessionObject[uid].settings;
        var sliderContainer = sessionObject[uid]["sliderContainer"];
        var numSlides = sessionObject[uid]["numSlides"];
        var slides = sessionObject[uid]["slides"];
        var dots = sessionObject[uid]["dots"];
        var index = sessionObject[uid]["currentIndex"];
        var prevIndex = sessionObject[uid]["prevIndex"];
        var oldIndex = index;
        index = checkSlideIndex(index, numSlides);

        sliderContainer.data("slider-index", index);
        sessionObject[uid].currentIndex = index;
        for (var i = 0; i < slides.length; i++) {
            $(dots[i]).removeClass("ofp-slider-dots-active");
            $(slides[i]).addClass("slide-invisible");
            $(slides[i]).removeClass("slide-visible");
            $(slides[i]).removeClass(settings.animation);
            $(slides[i]).removeClass(settings.animationRight);
            $(slides[i]).removeClass(settings.animationLeft);
        }

        if (settings.animate) {
            $(slides[index - 1]).removeClass("slide-invisible");
            if (settings.animation === "normal") {
                $(slides[index - 1]).css("opacity", 0.19);
                $(dots[index - 1]).addClass("ofp-slider-dots-active");
                $(slides[index - 1]).animate({
                    opacity: 1
                }, settings.duration, settings.easing, function () {
                    $(slides[index - 1]).addClass("slide-visible");
                });
            } else if (settings.animation === "css") {
                $(dots[index - 1]).addClass("ofp-slider-dots-active");
                if (prevIndex < oldIndex)
                    $(slides[index - 1]).addClass(settings.animationRight);
                else
                    $(slides[index - 1]).addClass(settings.animationLeft);
            }
        } else {
            $(dots[index - 1]).addClass("ofp-slider-dots-active");
            $(slides[index - 1]).removeClass("slide-invisible");
            $(slides[index - 1]).addClass("slide-visible");
        }
    }

    $.fn.offbeatSlider.settingsDefault = {
        slideStartIndex: 1,
        animate: true,
        animation: "normal",
        animationLeft: "",
        animationRight: "",
        duration: 1000,
        easing: "linear", //easein, linear, ...
        carousel: false, //auto-animate ?
        carouselDelay: 3000
    };

})(jQuery);
